### PR TITLE
key-state: make nr-pressed-keys check more accurate

### PR DESCRIPTION
**Pull request by johanmalm**
@jlindgren90 @Consolatis Maybe not the tidiest, but just a quick proof of concept to go with our discussion in issue #1208

### CODE REVIEW

File: src/key-state.c [hunk 1]
```diff
@@ -103,8 +103,38 @@ key_state_nr_bound_keys(void)
...
-	return pressed.nr_keys;
+	if (pressed.nr_keys < 2) {
+		return false;
+	}
+
+	/*
+	 * If the pressed array contains multiple keys, we have to analyze a bit
+	 * further because some keys like those listed below do not always
+	 * receive release events.
+	 *   - XKB_KEY_ISO_Group_Next
```
Comment (Consolatis) [inline src/key-state.c:121]: The key that sometimes is identified as `XKB_KEY_ISO_Group_Next` should always receive a release event.
This should be fixed by always removing the key on release events.

File: src/key-state.c [hunk 2]
```diff
@@ -103,8 +103,38 @@ key_state_nr_bound_keys(void)
...
+	 *   - XKB_KEY_XF86WakeUp
+	 *   - XKB_KEY_XF86Suspend
+	 *   - XKB_KEY_XF86Hibernate
+	 */
+	size_t count = 0;
+	for (int i = 0; i < pressed.nr_keys; ++i) {
+		/* Convert from udev event to xkbcommon code */
+		uint32_t keycode = pressed.keys[i] + 8;
+		if (keycode >= 0x1000000) {
```
Comment (Consolatis) [inline src/key-state.c:130]: AFAIR we are storing the actual keycode here, not the keysyms.
So if we'd want to decide based of the keysyms here we'd have to translate them first.

**Comment by Consolatis** [comment]
While this approach could fix the `keybinds-stuck` issue (for the sym ranges used) wouldn't it be an issue for the applications as well? AFAIR we are using the array (unless bound later) to send the initial state to clients when they get focus. Thus they see a `press` but never a release. I am not entirely sure here though.

**Comment by jlindgren90** [comment]
I still think it's worth at least considering dropping the check entirely. Having one (non-modifier) key still held doesn't prevent further key presses from being processed in most applications I'm familiar with. E.g. I tested typing "qw" into a terminal earlier, pressing "w" before releasing "q", and both letters appeared. I wouldn't personally expect the "w" to be ignored in that case. Not sure why labwc keybindings should be any different.

**Comment by jlindgren90** [comment]
Keeping the general "ignore multiple keys" rule while making exceptions for particular keys, sounds like it could turn into a game of whack-a-mole. My laptop has a WakeUp "key" that gets stuck. Some other system may have a different quirk and a different key gets stuck. Maybe someone spilled soda on their keyboard and one of the keys really does stick for a couple seconds - do we drop other key presses in that case?

**Comment by jlindgren90** [comment]
>  I guess the 2nd issue could be fixed by just storing the current (event->keycode) as bound (and removing it from 'pressed') rather than all the keys in the pressed set.

This makes sense to me -- we should not treat an unrelated pressed/stuck key as bound if it was not actually part of the keybinding.

> But that would make the first issue worse by sending the 'wakeup key pressed' (or any other stuck key) to all applications getting focus until restarting labwc.

I would expect applications to just ignore that key. Do we know of applications that don't?

I agree that everything we do here is managing symptoms -- the root cause in this case is probably a weird/buggy ACPI implementation resulting in that keypress without release.

**Comment by johanmalm** [comment]
Hmm.... some thinking time needed :smile:

**Comment by Consolatis** [comment]
Here might be another one:
- #1223

**Comment by jlindgren90** [comment]
> I guess the 2nd issue could be fixed by just storing the current (event->keycode) as bound (and removing it from 'pressed') rather than all the keys in the pressed set.

This part implemented in #1226.

**Pull request closed without merging**